### DISCUSSION TITLE
[Feature]: Return arbitrary Response object directly from loader

### DISCUSSION THREAD

**Discussion by colinclerk** [remix-run/remix#1880] [Proposals]
### What is the new or updated feature that you are suggesting?

We would like to return an arbitrary Response object directly from a loader.

### Why should this feature be included?

Our use case is that we're building an authentication library and we have special handling for the 401 Unauthorized case. We'd like to provide our handling directly through the loader, instead of asking developers to also modify the CatchBoundary.

Beyond our use case, it was a little confusing to learn that `redirect()` ~~or `json()`~~ Responses can be returned directly from a loader, but arbitrary Response objects cannot. We'd love a convention like returned Response objects are passed directly to the browser, while thrown Response objects go through the CatchBoundary.

**Comment by sergiodxa** [comment]
You can 100% return a new Response object with any status code and body from a loader.

If the route has a default export the data will be JSON.parse-d so your data should be compatible with that, for example don’t return a Buffer.

Also the json helper function can receive a status code or ResponseInit as second argument and there you can make the response created by it to be an unauthorized response.

**Comment by colinclerk** [comment]
Hi @sergiodxa - thanks for your help!

Our route has a default export and we want to return an HTML response directly from the loader under certain conditions.  We haven't been able to make this work - perhaps because of the JSON.parse restriction you mention?  If the capability already exists we'd love to use it!

Here's failing loader:
```js
export const loader: LoaderFunction = async () => {
  return new Response("<!DOCTYPE html><html><body>Test</body></html>", {
    status: 401,
  });
};
```

The error we receive is somewhat cryptic (it results in a 500 in the browser):
```
TypeError: Cannot read properties of undefined (reading 'toString')
    at Index (/Users/colinsidoti/my-remix-app/build/index.js:443:14)
    at processChild (/Users/colinsidoti/my-remix-app/node_modules/react-dom/cjs/react-dom-server.node.development.js:3353:14)
    at resolve (/Users/colinsidoti/my-remix-app/node_modules/react-dom/cjs/react-dom-server.node.development.js:3270:5)
    at ReactDOMServerRenderer.render (/Users/colinsidoti/my-remix-app/node_modules/react-dom/cjs/react-dom-server.node.development.js:3753:22)
    at ReactDOMServerRenderer.read (/Users/colinsidoti/my-remix-app/node_modules/react-dom/cjs/react-dom-server.node.development.js:3690:29)
    at renderToString (/Users/colinsidoti/my-remix-app/node_modules/react-dom/cjs/react-dom-server.node.development.js:4298:27)
    at handleRequest (/Users/colinsidoti/my-remix-app/build/index.js:389:51)
    at renderDocumentRequest (/Users/colinsidoti/my-remix-app/node_modules/@remix-run/server-runtime/server.js:404:18)
    at async requestHandler (/Users/colinsidoti/my-remix-app/node_modules/@remix-run/server-runtime/server.js:55:20)
    at async /Users/colinsidoti/my-remix-app/node_modules/@remix-run/express/server.js:45:22
```

**Comment by sergiodxa** [comment]
@colinclerk if the route has a default export the data of the loader is used to supply data for that route component, instead of sending html, send data and render that html directly in the component of the route.

Another option is to don’t have a default export so the route becomes a Resource Route.

**Comment by colinclerk** [comment]
Yes, this is where we're stuck.  We're building an SDK and don't want to ask the developer to modify both their default export and their loader.

We'd like to be able to return arbitrary Response objects from the loader, just like we can return redirect ~~and json~~ Response objects.

For context, we're able to do this from `getServerSideProps` in Next.js by calling:
```
ctx.res.writeHead(401, { 'Content-Type': 'text/html' });
ctx.res.end(body);
```

**Comment by colinclerk** [comment]
For anyone who came here through the PR - in the comments above I was originally confused about the behavior of the `json()` helper.  I thought a `json` response would skip the default export render, but in fact only `redirect()` skips the render.

**Comment by ryanflorence** [comment]
This is interesting though I'm not convinced yet 😅.  Loaders are designed for route components, not just arbitrary responses (resource routes can have arbitrary responses though). You have to consider both document requests as well as script transitions where multiple fetches are in flight to be aggregated into the next page in the browser.

I would give developers a component to render in their catch boundary for the 401 case.

**Reply by colinclerk** [reply]
It would definitely be a sharp knife!  We believe we have the script transition case handled, but can understand the hesitation.

Ultimately it looks like we'll be asking devs to modify **entry.server.tsx**.

**Reply by kiliman** [reply]
I actually had a use-case for this. While making the MDN Remix app, I have a route for documents that pulls in the data from MDN JSON endpoint. However, they also have images that are served up from the same folder. For example, the document `/en-US/docs/Learn/Getting_started_with_the_web/HTML_basics` also has an image `/en-US/docs/Learn/Getting_started_with_the_web/HTML_basics/grumpy-cat-small.png`.

In my loader, I was checking the extension and if it was an image, I would fetch the image from MDN. Initially I wanted to simply return the fetch response. But since this route had a default component, it was trying to render a non-JSON response.

I updated *entry.server* to check the `content-type` header and tried to return the response there. Unfortunately, the binary image data was already converted to a string by the time the entry was executed. I ended up converting the binary data to base64 and back.

I had another route just for `/static/*` assets (outside of a document), and these I simply used as a resource route and my loader returned the fetch response directly.

All this to say is that it would be nice if the loader sets the response `content-type` to something other than `application/json` that it doesn't try to stringify the body.

Like I said, I was able to work around it, so this is more of a "nice to have".

**Comment by filenwind** [comment]
I thought I need this feature to create graphql from remix-app-server with [graphql-helix](https://github.com/contra/graphql-helix)

The last step: `sendResult(result, res);`

I don't know how to send grapqhl response...`sendResult(result, new Response());` doesn't work.

Can someone help? Thanks!

**Reply by sergiodxa** [reply]
It looks like that library expects a Node response object, in that case, and if there’s no way to get the result instead of sending it directly, you will need to move the code to Express.

Check if there’s a way to get the result as an object instead of sending it.

**Reply by sergiodxa** [reply]
I think you will need to build a version of this https://github.com/contra/graphql-helix/blob/master/packages/core/lib/send-result/node-http.tsbut using Fetch Response instead of the Node Response

**Reply by filenwind** [reply]
Appreciate your help! I will take a try, thank you!

**Comment by seaglade** [comment]
I'm running into another use case for this right now. I'm implementing a system on my site to block specific clients. The logic for determining whether to block a given request works fine, but I want to return a `text/plain` response and unfortunately I can't seem to get it working. Currently I'm trying to implement a call in the `loader` of my `root.tsx` which determines whether a client is blocked and if so returns a Response that looks like this:
```ts
new Response('You are not allowed to access this site', {
	status: 403,
});
```
The 403 status seems to translate correctly when throwing that response from the loader, but the actual response body is HTML with "403" as text and a lot of CSS and JS that I don't want to be sending. Essentially I want the response to work the way that returning this from a resource route would work, but the existence of a default component seems to break it. Any ideas?

**Reply by seaglade** [reply]
For anyone who sees this, I solved it. Turns out it's not that hard: I just added this to my `entry.server.ts` in the `handleRequest` function before the existing return statement:
```ts
if (isBanned()) { // or any other condition you want
    return new Response('You are not allowed to access this site', { // or any other custom response data
        status: 403,
    });
}
```
~~The one caveat is that `handleRequest` is **not** `async`, which can be inconvenient depending on the logic you need to call into, but there are tricks to call `async` functions in `non-async` scopes so if you need to do that it's just hacky.~~ Adding `async` to the signature of `handleRequest` doesn't appear to cause any issues.
